src/interfaces/sampler_interface.jl

```diff
@@ -0,0 +1,140 @@
+#################
+# The sampler interface
+#
+# In this file, we define the interface for general samplers and specific
+# samplers for scattering processes.
+#################
+
+"""
+Abstract base type for general sampler.
+
+!!! note "Sampler interface"
+
+    Functions to be implemented
+
+    ```Julia
+    CustomSampler <: AbstractSampler{Float64}
+
+    Base.eltype  # type of the generated samples
+    QEDevents._weight
+    Base.size  # Return the dimensionality of the sampler (number of outgoing four-momenta per sample).
+    is_exact
+    QEDevents._rand!(rng::AbstractRNG, ::AbstractSampler, x::AbstractVector{T}) where {T}
+
+    ```
+    
+    optional
+
+    ```Julia
+    _post_processing
+    ```
+
+"""
+abstract type AbstractSampler <: AbstractComputationSetup end
+
+@inline Base.eltype(::AbstractSampler) = Float64
+
+"""
+Interface function, which asserts that the given `input` is valid.
+"""
+function _assert_valid_input(smplr::AbstractSampler, x::AbstractVecOrMat)
+    size(x, 1) == size(smplr, 1)[1] || throw(
+        InvalidInputError(
+            "The dimensionality of the input is $(size(x,1)) but it should be $(size(smplr,1)[1]).",
+        ),
+    )
+    return eltype(x) == eltype(smplr) || throw(
+        InvalidInputError(
+            "The element type of the input is $(eltype(x)) but it should be $(eltype(smplr)).",
+        ),
+    )
+end
+
+function QEDprocesses._compute(smplr::AbstractSampler, sample)
+    return _weight(smplr::AbstractSampler, sample)
+end
+
+"""
+
+    _weight(::AbstractSampler, sample)
+
+Interface function, which returns the weight associated with the given sample according to the sampler.
+
+!!! note ""
+    
+    This function must not do input validation. This is done by [`weight`](@ref) which calls `_weight` after input validation.
+
+"""
+function _weight end
+
+"""
+
+    setup(::AbstractSampler)
+
+Interface function, which returns the `setup::AbstractComputeSetup` associated with the sampler.
+"""
+function setup end
+
+"""
+
+    is_exact(::AbstractSampler)
+
+Interface function, which returns whether the sampler is exactly representing the base distribution or not.
+"""
+function is_exact end
+
+"""   
+
+    $(TYPEDSIGNATURES)
+
+Interface function, which validates the input, calculates the weight via [`_weight`](@ref) and performs an optional post-processing via QEDprocesses._post_processing.
+"""
+function weight(smplr::AbstractSampler, sample)
+    return compute(smplr, sample)
+end
+
+"""
+
+    _rand!(rng, ::AbstractSampler, ::AbstractVector{P}) where {P}
+
+Generate a random sample from the sampler and write it into the given vector.
+"""
+function _rand! end
```

Comment:
This is a type of piracy, since `Random._rand!` and `Distributions._rand!` are already implemented. I suggest removing this and dispatching one of the above (presumably from `Distributions`).

At some point, one could consider to directly interface with `Distributions.jl` via `Sampleable` and related functions (e.g. see https://github.com/JuliaStats/Distributions.jl/blob/e407fa5fd098e50df51801c6d062946eac7a7d0f/src/common.jl#L72).

Reply:
I removed the line in 2ad8208d75d466008955007b234b7999173f2368 so now it should be using Random._rand! since that was already imported in QEDevents.jl

Reply:
As of d4b5bb079f73aff9a8c1224c9dbdbfc67514f393 all variations of rand are imported from Distributions.jl